lib/models/event.js

```diff
@@ -142,5 +145,13 @@ module.exports.MatrixEvent.prototype = {
      */
     isState: function() {
         return this.event.state_key !== undefined;
+    },
+
+    getPushActions: function(client) {
```

Comment:
I was under the impression that we agreed that we should pass `client.pushRules` as an optional constructor parameter to `MatrixEvent`. On looking at the data you require for the push processor, I can see why you've done this though. This current signature feels wrong (passing it effectively the entire client instance) just to extract `client.pushRules`, `client.credentials.userId` and the `Room` instance further in. We should be explicit with our dependencies.

Given we require more things that just the push rules, this increasingly feels like we shouldn't be constructing and invoking `PushProcessor` in `MatrixEvent` at all. The reason we decided against doing this in `MatrixClient` was because it is relatively computationally expensive to do `PushProcessor.actionsForEvent`, given it will be called for *every* event. Perhaps we can pull out the `getPushActions` function to `MatrixClient`, and just invoke it directly on `MatrixEvent.getPushActions`? E.g.

```javascript
function MatrixEvent(event, pushActionsFn) {
  this.event = event;
  this._pushActionsFn = pushActionsFn;
}

MatrixEvent.prototype.getPushActions = function() {
  if (!this._pushActionsFn) {
    return null; // or whatever is suitable e.g. { }
  }
  return this._pushActionsFn(this.event);
```

And then have in `client.js`:

```javascript
// NB: Passing in 'room' because there isn't any guarantee that the store knows this room yet
//     (so client.getRoom(roomId) won't work)
function _PojoToMatrixEventMapper(client, room, plainOldJsObject) {
  var pushFn = function(event) {
    var pushProcessor = new PushProcessor(room, client.pushRules, client.credentials.userId);
    return pushProcessor.actionsForEvent(event);
  };
  return new MatrixEvent(plainOldJsObject, pushFn);
}
```

We would almost certainly want to optimise this (not creating a new `pushFn` function for every event ever; not creating a `new PushProcessor` every time if a static function would work just as well), but this is the general idea.

The benefits to this are:
 - Delegates push logic entirely out of `MatrixEvent` (similar to currently but currently you have to still supply the deps at the point of invocation rather than at construction time)
 - Makes `PushProcessor` dependencies explicit (much easier to test and get a high level overview of what is required in order to "process" correctly without having to comb through the code looking for method calls on `client`)

The drawback here is that it isn't entirely clear that `MatrixEvent.getPushActions` is just proxying through to a function `pushActionsFn` supplied in the constructor, but hey this is what JSDoc is for (and it also allows other people to inject in different ways of handling push rules if they're crazy enough).